features/bootstrap/FeatureContext.php

```diff
@@ -397,12 +397,17 @@ public function itShouldFail($success)
         $message = '';
 
         if ('fail' === $success) {
+            Assert::assertNotEquals(0, $this->getExitCode(), 'Step should FAIL but it PASS.');
+
             if (0 === $this->getExitCode()) {
                 $message = 'Actual output:' . PHP_EOL . PHP_EOL . $this->getOutput();
             }
 
             Assert::assertNotEquals(0, $this->getExitCode(), $message);
         } else {
+            Assert::assertEquals(0, $this->getExitCode(), 'Step should PASS but it FAIL.');
+
+
             if (0 !== $this->getExitCode()) {
                 $message = 'Actual output:' . PHP_EOL . PHP_EOL . $this->getOutput();
             }
```

Comment:
So this was changed quite recently in #1487 by @uuf6429 as with the previous behaviour it was hard to see *why* the scenario had failed without including the test output. I think this will reintroduce that problem.

I think a better solution might be to just throw our own exception here so that we can render the message ourselves in a more logical sequence.

Perhaps something like (haven't tested this):

```php
public function itShouldFail($success)
{
    $is_correct = match($success) {
        'fail' => 0 !== $this->getExitCode(),
        'pass' => 0 === $this->getExitCode(),
   }
 
   if ($is_correct) {
    return;
   }

   throw new UnexpectedValueException(
      'Expected previous command to '.$success.' but got exit code '.$this->getExitCode().PHP_EOL
      .PHP_EOL
      .'Actual command output: '.PHP_EOL
      .$this->getOutput()
    );
}
```

Reply:
Would need to look at the features in detail as explicitly asserting the output could be neater but my hunch is there are times when only a partial match on the output can be used, so having something that captures the full output could still be useful if the scenario fails unexpectedly (less so if it passes unexpectedly, probably).

Behat has no concept of Failed Vs Errored steps, and no specific domain for exceptions - it can't, because it relies entirely on the user supplying an assertion library (if they want one).
 https://docs.behat.org/en/latest/user_guide/context/definitions.html#failed-steps

Reply:
We could look into modifying what we do with the output but for now this solution would provide a more informative output, so I would probably just go ahead with it, either as it is or using the change proposed by @acoulton . Just one more thing, I would probably change the name of the function to `itShouldPassOrFail()` so that it is clearer that it's testing both possibilities

Reply:
That is, the proposed change could be simplified to just the following:
```php

    /**
     * Checks whether previously ran command failed|passed.
     *
     * @Then /^it should (fail|pass)$/
     *
     * @param 'fail'|'pass' $passOrFail
     */
    public function itShouldPassOrFail($passOrFail)
    {
        if ('fail' === $success) {
            Assert::assertNotEquals(0, $this->getExitCode(), 'Command should fail, but it succeeded.');
        } else {
            Assert::assertEquals(0, $this->getExitCode(), 'Command should succeed, but it failed.');
        }
    }
```
(I also fixed a few things, you're welcome. 😄)

Reply:
I would actually implement those as 2 separate steps instead of this `passOfFail` argument that switches the whole body of the method anyway

Reply:
Personally I would prefer to keep the output for debugging regardless of whether previous steps have done an assertion on partial output matches. There is nothing worse than getting a `it should have worked but it didn't` failure with nothing else to go on, IMO. Especially as in my experience that often happens either flickering or during dependency/platform updates.

If we had #1491 already this would be a good usecase for it. As we don't, I think we should just make sure the failure message is clear & first and the debug output follows it.

> I would actually implement those as 2 separate steps instead of this passOfFail argument that switches the whole body of the method anyway

Me too, but bear in mind this step is also called from other steps that follow the same pattern so those would have to be refactored too.

Reply:
I agree which is why I requested changes to this PR and suggested we just throw an exception ourselves so the output is clear.

Reply:
@uuf6429 @acoulton  I changed the output to be 

```
    When I run "behat --no-colors --format-settings='{\"paths\": false}' features/failing.feature" # FeatureContext::iRunBehat()
    Then it should pass with:                                                                      # FeatureContext::itShouldPassWith()
      """
      --- Failed scenarios:
      
          features/failing.feature:13
          features/failing.feature:18
      
      4 scenarios (2 passed, 2 failed)
      14 steps (10 passed, 2 failed, 2 skipped)
      """
      Expected previous command to PASS but got exit code 1
      
      ##### Actual output #####
      > Feature: Failing Feature
      >   In order to test the stop-on-failure feature
      >   As a behat developer
      >   I need to have a feature that fails
      > 
      >   Background:
      >     Given I have a step that passes
      > 
      >   Scenario: 1st Passing
      >     When I have a step that passes
      >     Then I should have a scenario that passed
      > 
      >   Scenario: 1st Failing
      >     When I have a step that passes
      >     And I have another step that fails
      >       step failed as supposed (Exception)
      >     Then I should have a scenario that failed
      > 
      >   Scenario: 2nd Failing
      >     When I have a step that fails
      >       step failed as supposed (Exception)
      >     Then I should have a scenario that failed
      > 
      >   Scenario: 2nd Passing
      >     When I have a step that passes
      >     And I have another step that passes
      >     Then I should have a scenario that passed
      > 
      > --- Failed scenarios:
      > 
      >     features/failing.feature:13
      >     features/failing.feature:18
      > 
      > 4 scenarios (2 passed, 2 failed)
      > 14 steps (10 passed, 2 failed, 2 skipped)
      ##### End of output ##### (UnexpectedValueException)
```

It's the best scenario I think, but I saw that there is two different steps `itShouldFail` and `itShouldFailWith`. If the output is not right, then the output does show only the diff in the output.